include/internal/thread.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OPENSSL_INTERNAL_WORKER_H
+# define OPENSSL_INTERNAL_WORKER_H
+# include <openssl/configuration.h>
+# if defined(OPENSSL_THREADS)
+
+#  include <openssl/types.h>
+#  include <openssl/crypto.h>
+#  include <internal/list.h>
+#  include <internal/cryptlib.h>
+#  include <internal/thread_arch.h>
+
+enum crypto_task_state {
+    CRYPTO_TASK_NO_STATE = 0,
+    CRYPTO_TASK_CREATED  = 1 << 0,
+    CRYPTO_TASK_STARTED  = 1 << 1,
+    CRYPTO_TASK_FINISHED = 1 << 2
+};
+
+typedef struct crypto_task_st {
+    uint64_t id;
+    void *data;
+    struct list list;
+
+    enum crypto_task_state state;
+
+    CRYPTO_THREAD_RETVAL retval;
+    CRYPTO_THREAD_ROUTINE routine;
+
+    /*
+     * CRYPTO_TASK::lock is taken:
+     *  - before task start advertisement is awaited (state transitions
+     *    from CRYPTO_TASK_STARTED, via CRYPTO_TASK::cond,
+     *    see crypto_thread_start_task)
+     *  - before task completion is awaited (state transition to
+     *    CRYPTO_TASK_FINISHED; via CRYPTO_TASK::cond,
+     *    see crypto_thread_join)
+     *  - before task's state is changed and broadcast,
+     *  - before task's retval is read
+     */
+    CRYPTO_MUTEX *lock;
+    CRYPTO_CONDVAR *cond;
+} CRYPTO_TASK;
+
+void *crypto_thread_start(OSSL_LIB_CTX *, CRYPTO_THREAD_ROUTINE, void *,
+                          uint8_t);
+void *crypto_thread_start_task(OSSL_LIB_CTX *ctx, CRYPTO_TASK *task,
+                               uint8_t options);
+int crypto_thread_join(OSSL_LIB_CTX *ctx, void *task,
+                       CRYPTO_THREAD_RETVAL *retval);
+size_t crypto_thread_get_available_threads(OSSL_LIB_CTX *ctx);
+int thread_spawn_worker_task(OSSL_LIB_CTX *ctx, CRYPTO_WORKER_CALLBACK cb,
+                             void *vtask);
+int crypto_task_clean(OSSL_LIB_CTX *ctx, void *task);
+int crypto_task_clean_all(OSSL_LIB_CTX *ctx);
```

Comment:
All these functions should have `ossl_` prefix.

Reply:
Resolved here: https://github.com/openssl/openssl/pull/12255/commits/0c644027aa27cd1acf8f28d721b3e368891b9eb0